sycl/test/reduction/reduction_nd_conditional.cpp

```diff
@@ -0,0 +1,131 @@
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUNx: env SYCL_DEVICE_TYPE=HOST %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+//==---reduction_nd_conditional.cpp - SYCL reduction + condition test ------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// This test performs basic checks of parallel_for(nd_range, reduction, func)
+// with reduction and conditional increment of the reduction variable.
+
+#include <CL/sycl.hpp>
+#include <cassert>
+
+using namespace cl::sycl;
+
+template <typename T, class BinaryOperation>
+void initInputData(buffer<T, 1> &InBuf, T &ExpectedOut, T Identity,
+                   BinaryOperation BOp, size_t N) {
+  ExpectedOut = Identity;
+  auto In = InBuf.template get_access<access::mode::write>();
+  for (int I = 0; I < N; ++I) {
+    if (std::is_same<BinaryOperation, std::multiplies<T>>::value)
+      In[I] = 1 + (((I % 37) == 0) ? 1 : 0);
+    else
+      In[I] = I + 1 + 1.1;
+
+    if (I < 2)
+      ExpectedOut = BOp(ExpectedOut, 99);
+    else if (I % 3)
+      ExpectedOut = BOp(ExpectedOut, In[I]);
+    else
+      ; // do nothing.
+  }
+};
+
+template <typename T, int Dim, class BinaryOperation>
+class Known;
+template <typename T, int Dim, class BinaryOperation>
+class Unknown;
+
+template <typename T>
```

Comment:
You could create a header file for definition of this class, since it is used in several tests

Reply:
Agreed.  I'd be tempted to rename it something like `CustomType` or `CustomVec` as well, to highlight why you're defining a new class instead of using `sycl::vec`.

Reply:
Ok, I created a new reduction_utils.hpp header and renamed the type. Thank you.